pyro/distributions/bivariate_von_mises.py

```diff
@@ -0,0 +1,257 @@
+
+    def sample(self, sample_shape=torch.Size()):
+        """ ** References: **
+              1. A New Unified Approach for the Simulation of aWide Class of Directional Distributions
+                John T. Kent, Asaad M. Ganeiber & Kanti V. Mardia (2018)
+        """
+
+        corr = self.correlation
+        conc = torch.stack((self.phi_concentration, self.psi_concentration))
+
+        eig = 0.5 * (conc[0] - corr ** 2 / conc[1])
+        eig = torch.stack((torch.zeros_like(eig), eig))
+        eigmin = torch.where(eig[1] < 0, eig[1], torch.zeros_like(eig[1], dtype=eig.dtype))
+        eig = eig - eigmin
+        eig.requires_grad = True
+        b0 = self._bfind(eig)
+
+        total = int(torch.prod(torch.tensor(sample_shape)))
+        missing = total * torch.ones(reduce(lambda a, b: a * b, self.batch_shape, 1), dtype=torch.int)
+        start = torch.zeros_like(missing)
+        phi = torch.empty((2, *missing.shape, total), dtype=corr.dtype)
+
+        max_iter = SineBivariateVonMises.max_sample_iter
+
+        # flatten batch_shape
+        conc = conc.view(2, -1, 1)
+        eigmin = eigmin.view(-1, 1)
+        corr = corr.view(-1, 1)
+        eig = eig.view(2, -1)
+        b0 = b0.view(-1)
+        phi_den = log_I1(0, conc[1]).view(-1, 1)
+
+        while torch.any(missing > 0) and max_iter:
+            curr_conc = conc[:, missing > 0, :]
+            curr_corr = corr[missing > 0]
+            curr_eig = eig[:, missing > 0]
+            curr_b0 = b0[missing > 0]
+
+            x = torch.distributions.Normal(0., torch.sqrt(1 + 2 * curr_eig / curr_b0)).sample(
+                (missing[missing > 0].min(),)).view(2, -1, missing[missing > 0].min())
+            x /= x.norm(dim=0)[None, ...]  # Angular Central Gaussian distribution
+
+            lf = curr_conc[0] * (x[0] - 1) + eigmin[missing > 0] + log_I1(0, torch.sqrt(
+                curr_conc[1] ** 2 + (curr_corr * x[1]) ** 2)).squeeze(0) - phi_den[missing > 0]
+            assert lf.shape == ((missing > 0).sum(), missing[missing > 0].min())
+
+            lg_inv = 1. - curr_b0.view(-1, 1) / 2 + torch.log(
+                curr_b0.view(-1, 1) / 2 + (curr_eig.view(2, -1, 1) * x ** 2).sum(0))
+            assert lg_inv.shape == lf.shape
+
+            accepted = torch.distributions.Uniform(0., 1.).sample(lf.shape) < (lf + lg_inv).exp()
+
+            phi_mask = torch.zeros((*missing.shape, total), dtype=torch.bool)
+            phi_mask[missing > 0] = torch.logical_and(
+                torch.arange(total).view(1, -1) < (start[missing > 0] + accepted.sum(-1)).view(-1, 1),
+                torch.arange(total).view(1, -1) >= start[missing > 0].view(-1, 1))
+
+            phi[:, phi_mask] = x[:, accepted]
+
+            start[missing > 0] += accepted.sum(-1)
+            missing[missing > 0] -= accepted.sum(-1)
+            max_iter -= 1
+
+        if max_iter == 0 or torch.any(missing > 0):
+            raise Exception("sample error")  # FIXME: what would be an appropriate exception?
```

Comment:
maybe a
```py
raise ValueError("maximum number of iterations exceeded; "
                 "try increasing `SineBivariateVonMises.max_sample_iter")
```

Reply:
Thanks for the suggestion; in my experience, when `max_iter` is hit, few to no samples are ever accepted, so the increase required is likely to be massive. 

I'll look into whether `max_iter` can be lower bounded or the growth can be estimated based on the distribution parameters, and maybe this estimate can be added to the error message.